x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1060,9 +1076,171 @@ private Expression propagate(And and) {
                         changed = true;
                     }
                 }
+
+                // evaluate all NotEquals against the Equal
+                for (Iterator<NotEquals> iter = frNotEquals.iterator(); iter.hasNext(); ) {
+                    NotEquals neq = iter.next();
+                    if (eq.left().semanticEquals(neq.left())) {
+                        Integer comp = BinaryComparison.compare(eq.right().fold(), neq.right().fold());
+                        if (comp != null) {
+                            if (comp == 0) {
+                                return FALSE; // clashing and conflicting: a = 1 AND a != 1
+                            } else {
+                                iter.remove(); // clashing and redundant: a = 1 AND a != 2
+                                changed = true;
+                            }
+                        }
+                    }
+                }
+
+                // evaluate all inequalities against the Equal
+                for (Iterator<BinaryComparison> iter = frInequalities.iterator(); iter.hasNext(); ) {
+                    BinaryComparison bc = iter.next();
+                    if (! eq.left().semanticEquals(bc.left())) {
+                        continue;
+                    }
+
+                    Integer compare = BinaryComparison.compare(eqValue, bc.right().fold());
+                    if (compare == null) {
+                        continue;
+                    }
+
+                    if (bc instanceof LessThan || bc instanceof LessThanOrEqual) { // a = 2 AND a </<= ?
+                        if ((compare == 0 && bc instanceof LessThan) || // a = 2 AND a < 2
+                            0 < compare) { // a = 2 AND a </<= 1
+                            return FALSE;
+                        }
+                    } else if (bc instanceof GreaterThan || bc instanceof GreaterThanOrEqual) { // a = 2 AND a >/>= ?
+                        if ((compare == 0 && bc instanceof GreaterThan) || // a = 2 AND a > 2
+                            compare < 0) { // a = 2 AND a >/>= 3
+                            return FALSE;
+                        }
+                    }
+
+                    iter.remove();
+                    changed = true;
+                }
+            }
+
+            return changed ? Predicates.combineAnd(CollectionUtils.combine(exps, frEquals, frNotEquals, frInequalities, ranges)) : and;
+        }
+
+        // combine disjunction:
+        // a = 2 OR a > 3 -> nop; a = 2 OR a > 1 -> a > 1
+        // a = 2 OR a < 3 -> a < 3; a = 2 OR a < 1 -> nop
+        // a = 2 OR 3 < a < 5 -> nop; a = 2 OR 1 < a < 3 -> 1 < a < 3; a = 2 OR 0 < a < 1 -> nop
+        // a = 2 OR a != 5 -> TRUE; a = 2 OR a = 5 -> nop
+        private Expression propagate(Or or) {
+            List<Expression> exps = new ArrayList<>();
+            List<Equals> frEquals = new ArrayList<>(); // foldable right term Equals
```

Comment:
I see that fr comes from foldable right - I would remove the prefix since there's no foldable left and the collections are used for compacting or selecting the expressions.

Reply:
Sure, removed it. I didn't like it either, tbh, but was trying to avoid the confusion as to why not all references of a certain type are part of the respective container.  But with the added comment (and code) it should be clear enough, indeed.